### PR TITLE
[Enhance] [Binlog] Reduce thread number of SyncJob to save resources

### LINKED ISSUE #6287: [Proposal] Support synchronizing MySQL binlog in real time
As we know, Binlog is the basic infrastructure in the Mysql Replication Architecture. The synchronization between replcations is carried out by reading and writing the binary log file (binlog) which stored on the mysql master server.

We know that in mysql cluster mode, only one replication is responsible for writing and the other replications are responsible for reading. Therefore, the repliactions' architecture is usually composed of one master (responsible for writing) and one or more slaves (responsible for reading). 

All data changes on the master node will be firstly write into local binlog, then copied to the slave nodes.

1. On the master node, the binlog file are named like mysql-bin.000001, mysql-bin.000002... MySQL will automatically segment binlog logs.

2. On the slave node, the binlog file name and position (offset) will be saved as a file or table to locate the latest consumption location. 

```
---------------------                                ---------------------
|       Slave       |                                |      Master       |
|                   |              read              |                   |
| FileName/Position | <<<--------------------------- |    Binlog Files   |
---------------------                                ---------------------
```

In order to get binlogs, the slave node sends the “MySQL binlog dump command” to the master node, and the dump thread of the master server will start to pushing binlog to the slave server continuously.

```
-----------                                             ------------
|  Slave  |                                             |  Master  |
-----------                                             ------------
     |                      1、连接                            |
     | --------------------------------------------------->>> |
     |                      2、握手                            |
     | <<<--------------------------------------------------- |
     |                      3、验证身份                         |
     | --------------------------------------------------->>> |
     |                      4、注册成功                         |
     | <<<----------------------------------------------------|
     |                      5、Mysql dump                     |
     | --------------------------------------------------->>> |
     
```

That is to say, we can get binlogs on the master node by forging this dump command. We can use Alibaba's Canal to achieve this goal.

Canal forge the dump protocol to disguise itself as a slave node to get and parse the master server's binlog log. Then it will store the parsed data in a ring queue in memory, waiting for other clients to subscribe and get it.

Therefore, with canal as the intermediary, Fe can get and synchronize the binlog logs on the master node. The blueprint for the first stage is below:

```
                                               ------------------------------------------------------          --------------      ------
                                               | ---------- ---------> channel 1 ---------> table1  |          |            | ---> | BE |
             Binlog                    Get     | | client | ---------> channel 2 ---------> table2  |          | Cordinator |      ------
---------- ---------> ----------- ------------>| ---------- ---------> channel 3 ---------> table3  | -------> |            |      ------
|  Mysql |            |  canal  |      Ack     |                                                    |          |     BE     | ---> | BE |
|        |            |         | <------------|  FE                                                |          |            |      ------
----------            -----------              ------------------------------------------------------          --------------
```

The work we need to do will be divided into two stages:

stage 1: 
1、Support creating consumption job and data channel in Fe to get parsed data in canal.
2、Support increasingly synchronizing the data changed in MYSQL, and ensure that the data isn't lost and repeateded.

stage 2:
1、Support synchronizing and executing MySQL DDL statements.
2、Embedding the canal into Fe, reduce one serialization, and no longer to independently deploy canal server.

### LINKED ISSUE #6417: [Enhance] [Binlog] Reduce thread number of sync job
SyncChannel is the data channel in SyncJob who is responsible for constantly delivering data for one single table ( See detail in #6287 ).

At present, for each channel our sync job allocates one thread to delivery data, which is very expensive. Especially when the number of channels become large, sync job will occupy a lot of CPU resources.

Therefore, we need to use thread pool to reduce the number of threads.

But the difficulty is how to keep the sent data in the correct order. We need to a special thread pool to ensure that the tasks belonging to the same channel are executed in the order of submission.

### PR DESCRIPTION
**Pull request by xy720**
## Proposed changes

#6417 

This commit is going to reduce thread number of SyncJob .
1、Submit send task to thread pool to send data.
2、Submit eof task to thread pool to block and wake up client to commit transactions.
3、Use SerialExecutorService to ensure correct order of sent data in every channel.

Besides，some bugs have been fixed in this commit
1、Failed to resume syncJob.
2、Failed to do sync data when set multiple tables in a syncJob.
3、In a cluster with multiple Fe, master may hang up after creating syncJob.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] Optimization. Including functional usability improvements and performance improvements.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #6287 #6417  ) and described the bug/feature there in detail

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/task/SyncTaskPool.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+package org.apache.doris.task;
+
+import java.util.concurrent.ExecutorService;
+
+public class SyncTaskPool {
+
+    private static final ExecutorService executor = new StripedTaskExecutor();
+
+    public SyncTaskPool() {
```
[the remark this answers is not in the document]
Comment (xy720) [reply fe/fe-core/src/main/java/org/apache/doris/task/SyncTaskPool.java]: Fixed

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 2]
```diff
@@ -634,6 +634,11 @@
      */
     @ConfField public static int sync_checker_interval_second = 5;
 
+    /**
+     * max num of thread to handle sync task in sync task thread-pool.
+     */
+    @ConfField public static int max_sync_task_threads_num = 1024;
```
[the remark this answers is not in the document]
Comment (xy720) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java]: Now is 10.

File: fe/fe-core/src/main/java/org/apache/doris/task/SerialExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                numOfSlots, 256, "sync-task-pool", true));
+    }
+
+    public void submit(Runnable command) {
+        int index = getIndex(command);
+        if (isSlotIndex(index)) {
+            SerialExecutor serialEx = slots[index];
+            serialEx.execute(command);
+        } else {
+            taskPool.execute(command);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/task/SerialExecutorService.java:58]: Do we need to accept task without index in this pool?

File: fe/fe-core/src/main/java/org/apache/doris/task/SyncTask.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.doris.task.SerialExecutorService.SerialRunnable;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public abstract class SyncTask implements SerialRunnable {
+    private static final Logger LOG = LogManager.getLogger(SyncTask.class);
+
+    protected long signature;
+    protected int index;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/task/SyncTask.java:30]: Add comment to explain the `index`

**Review by morningman — COMMENTED** [review]
Please add UT for `SyncTaskPool`

File: fe/fe-core/src/main/java/org/apache/doris/task/SyncTask.java (hunk 4 above)
Comment (xy720) [reply fe/fe-core/src/main/java/org/apache/doris/task/SyncTask.java:43]: done

File: fe/fe-core/src/main/java/org/apache/doris/task/SerialExecutorService.java (hunk 3 above)
Comment (xy720) [reply fe/fe-core/src/main/java/org/apache/doris/task/SerialExecutorService.java:58]: It‘s for compatibility with general thread pools

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ChannelDescription.java [hunk 5]
```diff
@@ -60,6 +60,8 @@
     // column names of source table
     @SerializedName(value = "colNames")
     private final List<String> colNames;
+    @SerializedName(value = "id")
```
[the remark this answers is not in the document]
Comment (xy720) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ChannelDescription.java]: done

**Review by morningman — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**